### ISSUE TITLE
better Boost test output

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#344]
**Reported by rcurtin on 12 Nov 44626608 04:21 UTC**
[libgdl](http://github.com/sumedhghaisas/libgdl/) by Sumedh implements an interesting wrapper around the Boost Unit Test Framework which makes the testing infrastructure gives output like

```
Running 26 test cases...
Starting test NormalParsingTest... [test MultipleArityErrorTest... [PASSED](PASSED]
Starting)
Starting test StratifiedNegationTest... [test StratifiedRecursionTest... [PASSED](PASSED]
Starting)
Starting test HeadSafetyTest... [test NotSafetyTest... [PASSED](PASSED]
Starting)
Starting test OrBoundTest... [test BaseDependencyTest... [PASSED](PASSED]
Starting)
Starting test InputDependencyTest... [test InitDependencyTest... [PASSED](PASSED]
Starting)
Starting test LeaglDependencyTest... [test RoleUndefinedTest... [PASSED](PASSED]
Starting)
Starting test LegalUndefinedTest... [test TerminalUndefinedTest... [PASSED](PASSED]
Starting)
Starting test GoalUndefinedTest... [test NormalReasonerTest... [PASSED](PASSED]
Starting)
Starting test QuestionTest... [test OrQuestionTest... [PASSED](PASSED]
Starting)
Starting test NotQuestionTest... [test DistinctQuestionTest... [PASSED](PASSED]
Starting)
Starting test RecursiveDependencyTest... [test EraseTest... [PASSED](PASSED]
Starting)
Starting test Question1Test... [test Question2Test... [PASSED](PASSED]
Starting)
Starting test CreateDSTest... [test LoadDSTest... [PASSED](PASSED]
Starting)
*** No errors detected
[100%] Built target test
```

This is done by the definition of a handful of useful macros in libgdl_test.hpp.  Sumedh and I think the same type of thing could be applied to the mlpack tests to clean up the output a lot, but there are a few intricacies:

 * It would be nice to not need to type MARK_START, MARK_PASSED, and all of the macro names into each test.  With how many tests are in mlpack this is somewhat unfeasible.  Maybe a MLPACK_AUTO_TEST_CASE macro should be defined that gives all of the relevant output.

 * The Log:: objects should have their output redirected to either stderr (probably not) or a file.  Otherwise the mlpack test output is drowned out by Log::Debug and Log::Warn messages.

 * The changes we make will have to be documented somewhere; it would probably be worth writing a quick tutorial so that users who are writing new tests for mlpack can understand the system we have set up.

Another thing that will help here is when I can make the Jenkins matrix build test across different versions of Boost, so that we can know that whatever we do will work on older (and newer) versions of Boost.Test...

Migrated-From: http://trac.research.cc.gatech.edu/fastlab/ticket/362

**Comment by rcurtin** [comment]
Basically, what this ticket comes down to is that it would be nice to print the currently running test for the sake of the user.  The Boost UTF might already do this, so this ticket might be very easy.

**Comment by apir8181** [comment]
I found something that might be useful from [Boost UTF Runtime Parameters](http://www.boost.org/doc/libs/1_57_0/libs/test/doc/html/utf/user-guide/runtime-config/reference.html#utf.user-guide.runtime-config.parameters).

After setting the parameter *log_level=test_suite*, Boost UTF print the currently running test information.

    Running 460 test cases...
    Entering test suite "MLPACKTest"
    Entering test suite "AdaBoostTest"
    Entering test case "HammingLossBoundIris"
    Leaving test case "HammingLossBoundIris"; testing time: 823709mks
    Entering test case "WeakLearnerErrorIris"
    Leaving test case "WeakLearnerErrorIris"; testing time: 840446mks

However, the Log:: objects still output to stdout, interfering UTF output. Logs objects probably should redirect their output to some places.

    Entering test case "DiscreteDistributionRandomTest" 
    [DEBUG]    1.4912e+04   3.0118e+04   4.9700e+03
    Leaving test case "DiscreteDistributionRandomTest"; testing time: 17148mks
    Entering test case "TestBooleanOption"
    [DEBUG] Compiled with debugging symbols.
    Leaving test case "TestBooleanOption"; testing time: 410mks

I want to tackle this problem, but I don't familiar with the framework of mlpack and tools it used(CMake, Boost UTF). Would you like to offer me a help?

**Comment by rcurtin** [comment]
Hi QiaoAn, thanks for looking into that.  The `Log` class in `src/mlpack/core/util/`, which is what `Log::Debug`, `Log::Info`, and `Log::Warn` are, can be disabled like so:

```Log::Debug.ignoreInput = true;```

But certainly we don't want to put three ignoreInput lines at the top of every test case.  Instead it may be better to modify the `Log` code so that ignoreInput defaults to true if a Boost test is being compiled (a user who is writing tests should be able to override this option, so that they can get output from `Log` if they want).

**Comment by rcurtin** [comment]
Fixed in PR #409.
